Mathlib/Algebra/Group/Submonoid/Membership.lean

```diff
@@ -274,6 +275,18 @@ theorem induction_of_closure_eq_top_right {s : Set M} {p : M → Prop} (hs : clo
   | one => exact H1
   | mul_right x _ y hy ih => exact Hmul x y hy ih
 
+@[to_additive]
+theorem exists_finset_prod_of_mem_closure {M : Type*} [CommMonoid M] {s : Set M} {x : M}
+    (hx : x ∈ closure s) :
+    ∃ (ι : Type) (I : Finset ι) (f : ι → M), (∀ i ∈ I, f i ∈ s) ∧ ∏ i ∈ I, f i = x := by
+  induction hx using Submonoid.closure_induction_left
+  case one                      => exact ⟨Empty, ∅, Empty.elim, by simp⟩
+  case mul_left y _ _ _ h_prod =>
+    rcases h_prod with ⟨_, I, f, _, rfl⟩;
+    exact ⟨Option I, Finset.univ, (Option.elim · y (f ·)),
+          ⟨by intro i; induction i <;> simp [*],
+          by simp [Finset.prod_attach]⟩⟩
```

Comment:
It usually is a bad idea to prove lemmas of the form "All elements in the closure of `s` are of form blah". Wherever you use those lemmas, there is a better proof using the closure induction principle, which is why those lemmas don't exist. Can you try rewriting your proofs without using this lemma?

Reply:
This lemma already exists in the list and multiset versions, and I'm applying it in this PR to prove directly the specialisation where `s := SumSqIn R`.
Would you like me to effectively move this proof into the proof of `IsSumSq.exists_sum`?
Alternatively, we could remove all 6 explicit sum lemmas from `SumsOfSquares.lean`. In that case, lemmas such as `IsSumSq.mul` would have to be proved by induction. However, I'm confused as I thought you previously indicated that you prefer the explicit sum-rewriting proof of `IsSumSq.mul` to the inductive one.

Reply:
I've just pushed a commit testing out the version without use of explicit sums (i.e., all done by induction). I'm happy with either; what do you (and others) think?

Reply:
Sorry, I guess I'm not internally consistent on the matter 🙈 Could you bring it up on Zulip to see what people think?

Reply:
RFC here:
https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/RFC.3A.20using.20vs.20avoiding.20explicit.20sum.20representations

Reply:
@urkud has pointed out to me that the issue is not relevant as the result I want (`IsSumSq.mul`) has already been proved in more generality. The discussion is still worth having because the same decision will need to be made regarding material I intend to PR very soon.

I'm rewriting the PR according to @urkud 's suggestion; this involves filling in some missing general lemmas and also making a new file. Will re-request a review in the coming days. Thanks for all the comments and discussion!